### ISSUE TITLE
Quiting Kodi gets delayed until EPG finished loading

### ISSUE THREAD

**Issue by EmilVogt** [xbmc/xbmc#15166]
### Describe the bug
To bo honest, I'm not sure if this is considered a bug, but I got really irritated the first time it happened. 
When having a slow internet connection and quiting Kodi right after startup, Kodi waits until the EPG is fully loaded before it actually quits. This can take (depending on your internet connection) more than 10 seconds and tricked me into thinking Kodi crashed. But waiting for long enough, Kodi finally closes. The problem is, when having a really really bad internet connection, this can get nasty.

## Expected Behavior
Kodi quits right away.

## Actual Behavior
Kodi waits until the EPG has finished downloading.

## Possible Fix
Abort downloading of EPG and quit Kodi.

### To Reproduce
1. Have slow internet connection
2. Start Kodi having the IPTV Simple Client running and set to load EPG data with every start of Kodi
3. Quit Kodi almost immediately before EPG finished downloading
4. Kodi will wait until EPG is fully loaded before it quits

### Debuglog
The debuglog can be found here:
[https://paste.kodi.tv/olapahusit.kodi](https://paste.kodi.tv/olapahusit.kodi)

## Additional context
I'm using kodinerds [.m3u-list](https://raw.githubusercontent.com/jnk22/kodinerds-iptv/master/iptv/kodi/kodi.m3u) and [this](https://github.com/sunsettrack4/xmltv_epg/raw/master/easyepg.gz) epg in the Simple IPTV client.

### Your Environment
Operating system version/name: Windows 10 (1803), 64bit
Kodi version: 18 (Leia) RC4, 64bit

**Comment by M-Reimer** [comment]
I have the same issue. I'm using VDR as PVR backend.
I'm not sure if the EPG is really the only thing that triggers this. My guess is, that loading the recording list also delays shutdown.

**Comment by ksooo** [comment]
Very valid points. Not only epg is affected, also recordings and timers loading can delay shutdown. Not easy to fix, unfortunatey.

**Comment by fuzzard** [comment]
Just a note for a little more info. A discussion internally about this

> optimized it, but yeah, Kodi waits on exit until unsafed EPG data are written. There are buggy PVR addons reporting each and every EPG tag changed every time so that Kodi considers all EPG data cahnged and if you have thousands of channels and a classic HDD on a slow NAS and Linux (reportedly), then this can take some time.
> 
> If you run Kodi more than just a couple of minutes (I call this developer mode, because normal users do not start Kodi many times in a row just to exit it without watching/listening anything), Kodi safes unsafed EPG data async in chunks while running.
> 
> If Kodi runs long enough and addons are not buggy there will be no or a very small amount of data on exit.

So, for a general use case, its believed this is a bit of a corner case that most users wont come across as they dont start/stop quickly. Some addons are incorrectly handling data, causing larger writes, and they would need to be fixed on the addon side.
With that said, its been requested to leave this open, as there is belief that some refinements could well be possible in core, so if anyone does look to dive into it, please do.

**Comment by ksooo** [comment]
Previous comment refers to another issue, which is about **storing** unsaved EPG tags on exit. This issue is about not exiting until all EPG data are **loaded**. Trechnically different, though not unrelated.
